File: app/models/user_profile.py
"""
User Profile Model with Goals, Preferences, and Activity Tracking
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class Gender(str, Enum):
    male = "male"
    female = "female"
    other = "other"


class ActivityLevel(str, Enum):
    sedentary = "sedentary"  # Little to no exercise
    lightly_active = "lightly_active"  # 1-3 days/week
    moderately_active = "moderately_active"  # 3-5 days/week
    very_active = "very_active"  # 6-7 days/week
    extremely_active = "extremely_active"  # Athlete level


class FitnessGoal(str, Enum):
    lose_weight = "lose_weight"
    maintain = "maintain"
    gain_muscle = "gain_muscle"
    improve_fitness = "improve_fitness"


class DietPreference(str, Enum):
    none = "none"
    vegetarian = "vegetarian"
    vegan = "vegan"
    pescatarian = "pescatarian"
    keto = "keto"
    paleo = "paleo"
    low_carb = "low_carb"
    high_protein = "high_protein"


class DailyGoals(BaseModel):
    """Daily nutrition and fitness goals"""
    calories: int = Field(..., description="Daily calorie target")
    protein_g: int = Field(..., description="Daily protein target (grams)")
    carbs_g: int = Field(..., description="Daily carbs target (grams)")
    fat_g: int = Field(..., description="Daily fat target (grams)")
    fiber_g: int = Field(default=25, description="Daily fiber target (grams)")
    water_ml: int = Field(default=2000, description="Daily water target (ml)")
    steps: int = Field(default=10000, description="Daily steps target")
    workouts_per_week: int = Field(default=3, description="Target workouts per week")


class UserProfile(BaseModel):
    """Complete user profile with goals and preferences"""
    user_id: str
    
    # Basic Info
    name: str
    gender: Gender
    age: int = Field(..., ge=13, le=120)
    height_cm: int = Field(..., ge=100, le=250)
    weight_kg: float = Field(..., ge=30, le=300)
    
    # Activity & Goals
    activity_level: ActivityLevel
    fitness_goal: FitnessGoal
    target_weight_kg: Optional[float] = None
    
    # Daily Goals
    daily_goals: DailyGoals
    
    # Preferences
    diet_preference: DietPreference = DietPreference.none
    allergies: List[str] = Field(default_factory=list)
    disliked_foods: List[str] = Field(default_factory=list)
    
    # Tracking
    onboarding_completed: bool = False
    current_streak: int = 0
    total_days_logged: int = 0
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Metadata
    timezone: str = "UTC"
    units: str = "metric"  # metric or imperial


def calculate_bmr(gender: Gender, weight_kg: float, height_cm: int, age: int) -> int:
    """
    Calculate Basal Metabolic Rate using Mifflin-St Jeor Equation
    BMR = calories burned at rest
    """
    if gender == Gender.male:
        bmr = (10 * weight_kg) + (6.25 * height_cm) - (5 * age) + 5
    else:  # female or other
        bmr = (10 * weight_kg) + (6.25 * height_cm) - (5 * age) - 161
    
    return int(bmr)


def calculate_tdee(bmr: int, activity_level: ActivityLevel) -> int:
    """
    Calculate Total Daily Energy Expenditure
    TDEE = BMR × Activity Multiplier
    """
    multipliers = {
        ActivityLevel.sedentary: 1.2,
        ActivityLevel.lightly_active: 1.375,
        ActivityLevel.moderately_active: 1.55,
        ActivityLevel.very_active: 1.725,
        ActivityLevel.extremely_active: 1.9,
    }
    
    return int(bmr * multipliers[activity_level])


def calculate_calorie_goal(tdee: int, fitness_goal: FitnessGoal) -> int:
    """
    Calculate daily calorie goal based on fitness objective
    """
    if fitness_goal == FitnessGoal.lose_weight:
        # 500 cal deficit for ~0.5kg/week loss
        return max(1200, tdee - 500)
    elif fitness_goal == FitnessGoal.gain_muscle:
        # 300 cal surplus for muscle gain
        return tdee + 300
    else:  # maintain or improve_fitness
        return tdee


def calculate_macro_goals(calorie_goal: int, fitness_goal: FitnessGoal, weight_kg: float) -> dict:
    """
    Calculate macro distribution based on goal
    Returns protein, carbs, fat in grams
    """
    if fitness_goal == FitnessGoal.lose_weight:
        # High protein (35%), moderate carbs (35%), moderate fat (30%)
        protein_g = int((calorie_goal * 0.35) / 4)  # 4 cal/g
        carbs_g = int((calorie_goal * 0.35) / 4)
        fat_g = int((calorie_goal * 0.30) / 9)  # 9 cal/g
    elif fitness_goal == FitnessGoal.gain_muscle:
        # Very high protein (40%), high carbs (40%), moderate fat (20%)
        protein_g = int((calorie_goal * 0.40) / 4)
        carbs_g = int((calorie_goal * 0.40) / 4)
        fat_g = int((calorie_goal * 0.20) / 9)
    else:  # maintain or improve_fitness
        # Balanced (30% protein, 40% carbs, 30% fat)
        protein_g = int((calorie_goal * 0.30) / 4)
        carbs_g = int((calorie_goal * 0.40) / 4)
        fat_g = int((calorie_goal * 0.30) / 9)
    
    # Ensure minimum protein (1.6g per kg bodyweight for active individuals)
    min_protein = int(weight_kg * 1.6)
    protein_g = max(protein_g, min_protein)
    
    return {
        "protein_g": protein_g,
        "carbs_g": carbs_g,
        "fat_g": fat_g,
        "fiber_g": 25,  # Standard recommendation
    }


def recommend_goals(
    gender: Gender,
    age: int,
    height_cm: int,
    weight_kg: float,
    activity_level: ActivityLevel,
    fitness_goal: FitnessGoal,
    target_weight_kg: Optional[float] = None
) -> DailyGoals:
    """
    Generate recommended daily goals based on user profile
    """
    bmr = calculate_bmr(gender, weight_kg, height_cm, age)
    tdee = calculate_tdee(bmr, activity_level)
    calorie_goal = calculate_calorie_goal(tdee, fitness_goal)
    macros = calculate_macro_goals(calorie_goal, fitness_goal, weight_kg)
    
    # Calculate water goal (30ml per kg bodyweight)
    water_ml = int(weight_kg * 30)
    
    # Workout recommendations
    workout_map = {
        FitnessGoal.lose_weight: 4,
        FitnessGoal.maintain: 3,
        FitnessGoal.gain_muscle: 5,
        FitnessGoal.improve_fitness: 4,
    }
    
    return DailyGoals(
        calories=calorie_goal,
        protein_g=macros["protein_g"],
        carbs_g=macros["carbs_g"],
        fat_g=macros["fat_g"],
        fiber_g=macros["fiber_g"],
        water_ml=water_ml,
        steps=10000,
        workouts_per_week=workout_map[fitness_goal],
    )


class ProfileUpdate(BaseModel):
    """Model for updating user profile"""
    name: Optional[str] = None
    gender: Optional[Gender] = None
    age: Optional[int] = Field(None, ge=13, le=120)
    height_cm: Optional[int] = Field(None, ge=100, le=250)
    weight_kg: Optional[float] = Field(None, ge=30, le=300)
    activity_level: Optional[ActivityLevel] = None
    fitness_goal: Optional[FitnessGoal] = None
    target_weight_kg: Optional[float] = None
    diet_preference: Optional[DietPreference] = None
    allergies: Optional[List[str]] = None
    disliked_foods: Optional[List[str]] = None
    daily_goals: Optional[DailyGoals] = None
    timezone: Optional[str] = None

File: flutter_app/lib/models/user_profile.dart
/// User Profile Model for Flutter
/// Matches backend UserProfile model
library;

enum Gender { male, female, other }

enum ActivityLevel {
  sedentary,
  lightlyActive,
  moderatelyActive,
  veryActive,
  extremelyActive
}

enum FitnessGoal { loseWeight, maintain, gainMuscle, improveFitness }

enum DietPreference {
  none,
  vegetarian,
  vegan,
  pescatarian,
  keto,
  paleo,
  lowCarb,
  highProtein
}

class DailyGoals {
  final int calories;
  final int proteinG;
  final int carbsG;
  final int fatG;
  final int fiberG;
  final int waterMl;
  final int steps;
  final int workoutsPerWeek;

  DailyGoals({
    required this.calories,
    required this.proteinG,
    required this.carbsG,
    required this.fatG,
    this.fiberG = 25,
    this.waterMl = 2000,
    this.steps = 10000,
    this.workoutsPerWeek = 3,
  });

  factory DailyGoals.fromJson(Map<String, dynamic> json) {
    return DailyGoals(
      calories: json['calories'] as int,
      proteinG: json['protein_g'] as int,
      carbsG: json['carbs_g'] as int,
      fatG: json['fat_g'] as int,
      fiberG: (json['fiber_g'] as int?) ?? 25,
      waterMl: (json['water_ml'] as int?) ?? 2000,
      steps: (json['steps'] as int?) ?? 10000,
      workoutsPerWeek: (json['workouts_per_week'] as int?) ?? 3,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'calories': calories,
      'protein_g': proteinG,
      'carbs_g': carbsG,
      'fat_g': fatG,
      'fiber_g': fiberG,
      'water_ml': waterMl,
      'steps': steps,
      'workouts_per_week': workoutsPerWeek,
    };
  }
}

class UserProfileModel {
  final String userId;
  final String name;
  final Gender gender;
  final int age;
  final int heightCm;
  final double weightKg;
  final ActivityLevel activityLevel;
  final FitnessGoal fitnessGoal;
  final double? targetWeightKg;
  final DailyGoals dailyGoals;
  final DietPreference dietPreference;
  final List<String> allergies;
  final List<String> dislikedFoods;
  final bool onboardingCompleted;
  final int currentStreak;
  final int totalDaysLogged;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String timezone;
  final String units;
  final String subscriptionTier;

  UserProfileModel({
    required this.userId,
    required this.name,
    required this.gender,
    required this.age,
    required this.heightCm,
    required this.weightKg,
    required this.activityLevel,
    required this.fitnessGoal,
    this.targetWeightKg,
    required this.dailyGoals,
    this.dietPreference = DietPreference.none,
    this.allergies = const [],
    this.dislikedFoods = const [],
    this.onboardingCompleted = false,
    this.currentStreak = 0,
    this.totalDaysLogged = 0,
    required this.createdAt,
    required this.updatedAt,
    this.timezone = 'UTC',
    this.units = 'metric',
    this.subscriptionTier = 'free',
  });

  factory UserProfileModel.fromJson(Map<String, dynamic> json) {
    return UserProfileModel(
      userId: json['user_id'] as String,
      name: json['name'] as String,
      gender: _genderFromString(json['gender'] as String),
      age: json['age'] as int,
      heightCm: json['height_cm'] as int,
      weightKg: (json['weight_kg'] as num).toDouble(),
      activityLevel: _activityLevelFromString(json['activity_level'] as String),
      fitnessGoal: _fitnessGoalFromString(json['fitness_goal'] as String),
      targetWeightKg: json['target_weight_kg'] != null
          ? (json['target_weight_kg'] as num).toDouble()
          : null,
      dailyGoals: DailyGoals.fromJson(json['daily_goals'] as Map<String, dynamic>),
      dietPreference: _dietPreferenceFromString(json['diet_preference'] as String? ?? 'none'),
      allergies: (json['allergies'] as List<dynamic>?)?.cast<String>() ?? [],
      dislikedFoods: (json['disliked_foods'] as List<dynamic>?)?.cast<String>() ?? [],
      onboardingCompleted: json['onboarding_completed'] as bool? ?? false,
      currentStreak: json['current_streak'] as int? ?? 0,
      totalDaysLogged: json['total_days_logged'] as int? ?? 0,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      timezone: json['timezone'] as String? ?? 'UTC',
      units: json['units'] as String? ?? 'metric',
      subscriptionTier: json['subscription_tier'] as String? ?? 'free',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'name': name,
      'gender': _genderToString(gender),
      'age': age,
      'height_cm': heightCm,
      'weight_kg': weightKg,
      'activity_level': _activityLevelToString(activityLevel),
      'fitness_goal': _fitnessGoalToString(fitnessGoal),
      'target_weight_kg': targetWeightKg,
      'daily_goals': dailyGoals.toJson(),
      'diet_preference': _dietPreferenceToString(dietPreference),
      'allergies': allergies,
      'disliked_foods': dislikedFoods,
      'onboarding_completed': onboardingCompleted,
      'current_streak': currentStreak,
      'total_days_logged': totalDaysLogged,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'timezone': timezone,
      'units': units,
      'subscription_tier': subscriptionTier,
    };
  }

  static Gender _genderFromString(String value) {
    switch (value.toLowerCase()) {
      case 'male':
        return Gender.male;
      case 'female':
        return Gender.female;
      default:
        return Gender.other;
    }
  }

  static String _genderToString(Gender gender) {
    return gender.name;
  }

  static ActivityLevel _activityLevelFromString(String value) {
    switch (value) {
      case 'sedentary':
        return ActivityLevel.sedentary;
      case 'lightly_active':
        return ActivityLevel.lightlyActive;
      case 'moderately_active':
        return ActivityLevel.moderatelyActive;
      case 'very_active':
        return ActivityLevel.veryActive;
      case 'extremely_active':
        return ActivityLevel.extremelyActive;
      default:
        return ActivityLevel.moderatelyActive;
    }
  }

  static String _activityLevelToString(ActivityLevel level) {
    switch (level) {
      case ActivityLevel.sedentary:
        return 'sedentary';
      case ActivityLevel.lightlyActive:
        return 'lightly_active';
      case ActivityLevel.moderatelyActive:
        return 'moderately_active';
      case ActivityLevel.veryActive:
        return 'very_active';
      case ActivityLevel.extremelyActive:
        return 'extremely_active';
    }
  }

  static FitnessGoal _fitnessGoalFromString(String value) {
    switch (value) {
      case 'lose_weight':
        return FitnessGoal.loseWeight;
      case 'maintain':
        return FitnessGoal.maintain;
      case 'gain_muscle':
        return FitnessGoal.gainMuscle;
      case 'improve_fitness':
        return FitnessGoal.improveFitness;
      default:
        return FitnessGoal.maintain;
    }
  }

  static String _fitnessGoalToString(FitnessGoal goal) {
    switch (goal) {
      case FitnessGoal.loseWeight:
        return 'lose_weight';
      case FitnessGoal.maintain:
        return 'maintain';
      case FitnessGoal.gainMuscle:
        return 'gain_muscle';
      case FitnessGoal.improveFitness:
        return 'improve_fitness';
    }
  }

  static DietPreference _dietPreferenceFromString(String value) {
    switch (value) {
      case 'vegetarian':
        return DietPreference.vegetarian;
      case 'vegan':
        return DietPreference.vegan;
      case 'pescatarian':
        return DietPreference.pescatarian;
      case 'keto':
        return DietPreference.keto;
      case 'paleo':
        return DietPreference.paleo;
      case 'low_carb':
        return DietPreference.lowCarb;
      case 'high_protein':
        return DietPreference.highProtein;
      default:
        return DietPreference.none;
    }
  }

  static String _dietPreferenceToString(DietPreference pref) {
    switch (pref) {
      case DietPreference.none:
        return 'none';
      case DietPreference.vegetarian:
        return 'vegetarian';
      case DietPreference.vegan:
        return 'vegan';
      case DietPreference.pescatarian:
        return 'pescatarian';
      case DietPreference.keto:
        return 'keto';
      case DietPreference.paleo:
        return 'paleo';
      case DietPreference.lowCarb:
        return 'low_carb';
      case DietPreference.highProtein:
        return 'high_protein';
    }
  }
}

// Helper extensions for UI display
extension GenderExtension on Gender {
  String get displayName {
    switch (this) {
      case Gender.male:
        return 'Male';
      case Gender.female:
        return 'Female';
      case Gender.other:
        return 'Other';
    }
  }

  String get emoji {
    switch (this) {
      case Gender.male:
        return '👨';
      case Gender.female:
        return '👩';
      case Gender.other:
        return '🧑';
    }
  }
}

extension ActivityLevelExtension on ActivityLevel {
  String get displayName {
    switch (this) {
      case ActivityLevel.sedentary:
        return 'Sedentary';
      case ActivityLevel.lightlyActive:
        return 'Lightly Active';
      case ActivityLevel.moderatelyActive:
        return 'Moderately Active';
      case ActivityLevel.veryActive:
        return 'Very Active';
      case ActivityLevel.extremelyActive:
        return 'Extremely Active';
    }
  }

  String get description {
    switch (this) {
      case ActivityLevel.sedentary:
        return 'Little to no exercise';
      case ActivityLevel.lightlyActive:
        return '1-3 days/week';
      case ActivityLevel.moderatelyActive:
        return '3-5 days/week';
      case ActivityLevel.veryActive:
        return '6-7 days/week';
      case ActivityLevel.extremelyActive:
        return 'Athlete level';
    }
  }

  String get emoji {
    switch (this) {
      case ActivityLevel.sedentary:
        return '🪑';
      case ActivityLevel.lightlyActive:
        return '🚶';
      case ActivityLevel.moderatelyActive:
        return '🏃';
      case ActivityLevel.veryActive:
        return '💪';
      case ActivityLevel.extremelyActive:
        return '🏋️';
    }
  }
}

extension FitnessGoalExtension on FitnessGoal {
  String get displayName {
    switch (this) {
      case FitnessGoal.loseWeight:
        return 'Lose Weight';
      case FitnessGoal.maintain:
        return 'Maintain Weight';
      case FitnessGoal.gainMuscle:
        return 'Gain Muscle';
      case FitnessGoal.improveFitness:
        return 'Improve Fitness';
    }
  }

  String get description {
    switch (this) {
      case FitnessGoal.loseWeight:
        return 'Lose fat while preserving muscle';
      case FitnessGoal.maintain:
        return 'Stay at current weight';
      case FitnessGoal.gainMuscle:
        return 'Build muscle mass';
      case FitnessGoal.improveFitness:
        return 'Get healthier and stronger';
    }
  }

  String get emoji {
    switch (this) {
      case FitnessGoal.loseWeight:
        return '📉';
      case FitnessGoal.maintain:
        return '⚖️';
      case FitnessGoal.gainMuscle:
        return '💪';
      case FitnessGoal.improveFitness:
        return '🎯';
    }
  }
}

extension DietPreferenceExtension on DietPreference {
  String get displayName {
    switch (this) {
      case DietPreference.none:
        return 'No Preference';
      case DietPreference.vegetarian:
        return 'Vegetarian';
      case DietPreference.vegan:
        return 'Vegan';
      case DietPreference.pescatarian:
        return 'Pescatarian';
      case DietPreference.keto:
        return 'Keto';
      case DietPreference.paleo:
        return 'Paleo';
      case DietPreference.lowCarb:
        return 'Low Carb';
      case DietPreference.highProtein:
        return 'High Protein';
    }
  }
}






